tests/functional/test_configured_endpoint_url.py

```diff
@@ -0,0 +1,283 @@
+import json
+from pathlib import Path
+from unittest import mock
+
+import pytest
+
+import botocore.configprovider
+import botocore.utils
+from botocore.compat import urlsplit
+from tests import ClientHTTPStubber, temporary_file
+
+ENDPOINT_TESTDATA_FILE = (
+    Path(__file__).parent
+    / 'data'
+    / "configured-endpoint-urls"
+    / "test-cases"
+    / "profile-tests.json"
+)
+
+with open(ENDPOINT_TESTDATA_FILE) as f:
+    data = json.load(f)
+    suites = data.get('testSuites', {})
+    ENDPOINT_TEST_SUITE = suites[0]
+
+
+def dict_to_ini_section(ini_dict, section_header):
+    profile_str = f'[{section_header}]\n'
+    for key, value in ini_dict.items():
+        if isinstance(value, dict):
+            profile_str += f"{key} =\n"
+            for new_key, new_value in value.items():
+                profile_str += f"  {new_key}={new_value}\n"
+        else:
+            profile_str += f"{key}={value}\n"
+    return profile_str + "\n"
+
+
+def create_cases(test_suite):
+    profiles = test_suite['profiles']
+    service_sections = test_suite['services']
+    environments = test_suite['environments']
+    client_configs = test_suite['client_configs']
+    tests = test_suite['endpointUrlTests']
+
+    for test in tests:
+        new_test = test.copy()
+
+        augmented_data = {}
+        environment_name = test.get('environment', None)
+        environment = environments.get(environment_name, {})
+        augmented_data['environment'] = environment
+
+        client_config_name = test.get('client_config', None)
+        client_config = client_configs.get(client_config_name, {})
+        augmented_data['client_config'] = client_config
+
+        profile_name = test.get('profile', None)
+        profile = profiles.get(profile_name, {})
+        augmented_data['profile'] = profile
+
+        service_section_name = profile.get('services', None)
+        service_section = service_sections.get(service_section_name, {})
+        augmented_data['services'] = service_section
+
+        new_test['augmented_data'] = augmented_data
+        yield new_test
+
+
+def _normalize_endpoint(url):
+    split_endpoint = urlsplit(url)
+    actual_endpoint = f"{split_endpoint.scheme}://{split_endpoint.netloc}"
+    return actual_endpoint
+
+
+def assert_endpoint_url_used(client, expected_endpoint_url):
```

Comment:
Let's rename this to something more explicit to make it more clear on the type of assertion we are doing. Mainly based on the assertion name here, I'd assume it would be making an assertion similar to `assert_endpoint_url_used_for_operation`. Maybe we can instead call it `assert_client_endpoint_url` to indicate we are asserting the endpoint_url on the client?